File: src/main/java/com/github/jeffreystoke/winsock/io/internal/WinSock.java
package com.github.jeffreystoke.winsock.io.internal;

import com.github.jeffreystoke.winsock.io.util.NativeUtils;

import java.io.IOException;


public class WinSock {

    public static native String _wsa_startup();

    static {
        try {
//            System.load("C:/Users/Jeffrey/Desktop/workshop/winsock-jni/build/libs/winsock_jni/shared/windows_amd64/winsock_jni.dll");
            System.loadLibrary("winsock_jni");
        } catch (UnsatisfiedLinkError e) {
            String platform = "windows_";
            String arch = System.getProperty("os.arch");
            if ("x86".equalsIgnoreCase(arch) || "i386".equalsIgnoreCase(arch) ||
                    "i486".equalsIgnoreCase(arch) || "i586".equalsIgnoreCase(arch) ||
                    "i686".equalsIgnoreCase(arch)) {
                platform += "i386";
            } else if ("x86_64".equalsIgnoreCase(arch) || "amd64".equalsIgnoreCase(arch)) {
                platform += "amd64";
            } else {
                throw new UnsupportedOperationException("unsupported arch");
            }

            try {
                NativeUtils.INSTANCE.loadLibraryFromJar("/natives/" + platform + "/winsock_jni.dll");
            } catch (IOException e1) {
                e1.printStackTrace();
            }
        }

        String err = _wsa_startup();
        if (err != null) {
            throw new RuntimeException(err);
        }
    }

    public static native void _free(long ptr);

    public static native void _wsa_cleanup();

    public static native int _wsa_get_last_error();

    public static native int _setsockopt(long s, int level, int optKey, byte[] optVal);

    public static native int _getsockopt(long s, int level, int optKey, byte[] optVal);

    public static native long _ioctlsocket(long s, boolean nonBlock);

    public static native long _socket(int addressFamily, int socketType, int proto);

    public static native long _bind(long s, String address, int port);

    public static native int _listen(long s, int backlog);

    public static native long _accept(long s, byte[] addressBuf, byte[] portBuf);

    public static native int _recv(long s, byte[] buf, int flag);

    public static native int _send(long s, byte[] buf, int flag);

    public static native int _closesocket(long s);

    public static native int _shutdown(long s, int how);

    public static native int _connect(long s, String address, int port);

    public static native int _recv_from(long s, byte[] buf, int flag, byte[] addressBuf, byte[] portBuf);

    public static native int _send_to(long s, byte[] buf, int flag, String remoteAddress, int port);

    public static native long _create_fd_set();

    public static native void _add_fd(long fdSet, long s);

    public static native void _remove_fd(long fdSet, long s);

    public static native long _get_set_fd(long fdSet);

    public static native int _select(long readFDs, long writeFDs, long exceptionFDs, int timeout);

    public static native long _wsa_create_event();

    public static native boolean _wsa_reset_event(long event);

    public static native boolean _wsa_close_event(long event);

    public static native int _wsa_wait_for_multiple_events(int count, long[] events, boolean waitAll, int waitTimeout, boolean alert_able);

    public static native int _wsa_event_select(long s, long event, long netEvents);

    public static native int _wsa_enum_network_events(long s, long event, byte[] bufForLong, byte[] bufForInt);

    public static native long _wsa_socket(int addressFamily, int socketType, int proto, int flag);

    public static native int _wsa_recv(long s, long[] buffers, int bytesToRead, int flag, long overlappedStruct);

    public static native int _wsa_send(long s, long[] buffers, int bytesToSend, int flag, long overlappedStruct);

    public static native long _create_wsa_buf(int size);

    public static native void _destroy_wsa_buf(long wsaBuf);

    public static native long _create_completion_key(long socket, String remoteAddress, int port);

    public static native long _create_wsa_overlapped(int internal, int internalHigh, int offset, int offsetHigh, long wsaEvent);

    public static native boolean _wsa_get_overlapped_result(long s, long wsaOverlapped, int bytesToTrans, boolean wait, long flag);

    public static native long _create_iocp(long fileHandle, long existingCP, long completionKey, int numOfCurrentThreads);

    public static native boolean _get_queued_completion_status(long cpHandler, int bytesToTrans, long completionKey, long overlapped, int waitTimeout);

    public static native boolean _cancel_io(long fileHandle);

    public static native boolean _post_queued_completion_status(long cpHandler, int bytesToTrans, long completionKey, long overlapped);
}

File: src/main/java/com/github/jeffreystoke/winsock/io/internal/WinSock.kt
@file:Suppress("FunctionName")

package com.github.jeffreystoke.winsock.io.internal

typealias Pointer = Long

object WinSock {

    init {
        System.loadLibrary("winsock_jni")
        _wsa_startup()?.let { err -> throw RuntimeException(err) }
    }

    @JvmStatic
    external fun _free(ptr: Pointer)

    @JvmStatic
    private external fun _wsa_startup(): String?

    @JvmStatic
    external fun _wsa_cleanup()

    @JvmStatic
    external fun _wsa_get_last_error(): Int

    @JvmStatic
    external fun _setsockopt(s: Pointer, level: Int, optKey: Int, optVal: ByteArray): Int

    @JvmStatic
    external fun _getsockopt(s: Pointer, level: Int, optKey: Int, optVal: ByteArray): Int

    @JvmStatic
    external fun _ioctlsocket(s: Pointer, mode: Int)

    @JvmStatic
    external fun _socket(addressFamily: Int, socketType: Int, proto: Int): Pointer

    @JvmStatic
    external fun _bind(s: Pointer, address: String, port: Int)

    @JvmStatic
    external fun _listen(s: Pointer, backlog: Int): Int

    // TODO 远端地址记录
    @JvmStatic
    external fun _accept(s: Pointer): Pointer

    @JvmStatic
    external fun _recv(s: Pointer, buf: ByteArray, flag: Int): Int

    @JvmStatic
    external fun _send(s: Pointer, buf: ByteArray, flag: Int): Int

    @JvmStatic
    external fun _closesocket(s: Pointer): Int

    @JvmStatic
    external fun _shutdown(s: Pointer, how: Int): Int

    @JvmStatic
    external fun _connect(s: Pointer, address: String, port: Int): Int

    // UDP
    // TODO 远端地址记录
    @JvmStatic
    external fun _recvFrom(s: Pointer, buf: ByteArray, flag: Int): Int

    @JvmStatic
    external fun _sendTo(s: Pointer, buf: ByteArray, flag: Int, remoteAddress: String): Int

    @JvmStatic
    external fun _create_fd_set(): Pointer

    @JvmStatic
    external fun _add_fd(fdSet: Pointer, s: Pointer)

    @JvmStatic
    external fun _remove_fd(fdSet: Pointer, s: Pointer)

    @JvmStatic
    external fun _get_set_fd(fdSet: Pointer): Pointer

    @JvmStatic
    external fun _select(readFDs: Pointer, writeFDs: Pointer, exceptionFDs: Pointer, timeout: Int): Int

    // 事件通知
    @JvmStatic
    external fun _wsa_create_event(): Pointer

    @JvmStatic
    external fun _wsa_reset_event(event: Pointer): Boolean

    @JvmStatic
    external fun _wsa_close_event(event: Pointer): Boolean

    @JvmStatic
    external fun _wsa_wait_for_multiple_events(count: Int, events: Array<Pointer>, waitAll: Boolean, waitTimeout: Int, alert_able: Boolean): Int

    @JvmStatic
    external fun _wsa_event_select(s: Pointer, event: Pointer, netEvents: Long): Int

    @JvmStatic
    external fun _wsa_enum_network_events(s: Pointer, event: Pointer, bufForLong: ByteArray, bufForInt: ByteArray): Int

    @JvmStatic
    external fun _wsa_connect()

    @JvmStatic
    external fun _wsa_accept(s: Pointer): Pointer

    // 重叠 I/O
    @JvmStatic
    external fun _wsa_socket(addressFamily: Int, socketType: Int, proto: Int, flag: Int): Pointer

    @JvmStatic
    external fun _wsa_recv(s: Pointer, buf: Array<ByteArray>, toRead: Int, flag: Int, overlappedStruct: Pointer)

    @JvmStatic
    external fun _wsa_recv_from()

    @JvmStatic
    external fun _wsa_send()

    @JvmStatic
    external fun _wsa_send_to()

    @JvmStatic
    external fun _create_completion_key(socket: Pointer, clientAddress: String): Pointer

    @JvmStatic
    private external fun _create_wsa_overlapped(internal: Int, internalHigh: Int, offset: Int, offsetHigh: Int, wsaEvent: Pointer): Pointer

    @JvmStatic
    external fun _wsa_get_overlapped_result(s: Pointer, wsaOverlapped: Pointer, wait: Boolean, flag: Pointer): Boolean

    // IOCP
    @JvmStatic
    external fun _create_iocp(fileHandle: Pointer, existingCP: Pointer, completionKey: Pointer, numOfCurrentThreads: Int): Pointer

    @JvmStatic
    external fun _get_queued_completion_status(cpHandler: Pointer, bytesRead: Int, completionKey: Pointer, overlapped: Pointer, waitTimeout: Int): Boolean

    @JvmStatic
    external fun _cancel_io(fileHandle: Pointer)

    @JvmStatic
    external fun _post_queued_completion_status(cpHandler: Pointer, bytesTx: Int, completionKey: Pointer, overlapped: Pointer): Boolean
}
